### PR TITLE
[MM-17442] Visual verification of "Searching" animation for Flagged and Pinned posts

### PR DESCRIPTION
**Pull request by sij507**
Add E2E test for M17442 Visual verification of "Searching" animation for Flagged and Pinned posts
Ticket: https://github.com/mattermost/mattermost-server/issues/11741

### CODE REVIEW

File: e2e/cypress/integration/post_header/post_header_spec.js [hunk 1]
```diff
@@ -217,4 +217,21 @@ describe('Post Header', () => {
...
+
+    it('M17442 Visual verification of "Searching" animation for Flagged and Pinned posts', () => {
+        // # Go to Town-Square channel
+        cy.visit('/ad-1/channels/town-square');
+
+        // # Click on the "Pinned Posts" icon to the left of the "Search" box
+        cy.get('#channelHeaderPinButton').click();
+
+        // * Check before pinned posts are loaded, a "Searching..." animation appears on the top of the RHS
+        cy.get('#loadingSpinner').should('be.visible').and('have.text', 'Searching...');
```
Comment (saturninoabril) [inline e2e/cypress/integration/post_header/post_header_spec.js]: Please have this verification within the RHS. That could be:
```javascript
        // * Verify that the RHS for pinned posts is opened.
        cy.get('#searchContainer').should('be.visible').within(() => {
            // * Check that searching indicator appears before the pinned posts are loaded
            cy.get('#loadingSpinner').should('be.visible').and('have.text', 'Searching...');
            cy.get('#search-items-container').should('be.visible');

            // # Close the RHS
            cy.get('#searchResultsCloseButton').should('be.visible').click();
        });
```

Note: Should add new searchContainer ID [here](https://github.com/mattermost/mattermost-webapp/blob/master/components/search_results/search_results.jsx#L293).

File: e2e/cypress/integration/post_header/post_header_spec.js [hunk 2]
```diff
@@ -217,4 +217,21 @@ describe('Post Header', () => {
...
+
+        // # Click on the "Flagged Posts" icon to the right of the "Search" box
+        cy.get('#channelHeaderFlagButton').click();
+
+        // * Check before flagged posts are loaded, a "Searching..." animation appears on the top of the RHS
+        cy.get('#loadingSpinner').should('be.visible').and('have.text', 'Searching...');
```
Comment (saturninoabril) [inline e2e/cypress/integration/post_header/post_header_spec.js]: Same here.

File: e2e/cypress/integration/post_header/post_header_spec.js (hunk 1 above)
Comment (sij507) [reply e2e/cypress/integration/post_header/post_header_spec.js]: @saturninoabril I got this error message
Screen Shot 2019-10-28 at 11 51 27 PM. I am not sure where to add the new searchContainer ID in the file you provided. Can you give me an example? Thanks!

Comment (saturninoabril) [reply e2e/cypress/integration/post_header/post_header_spec.js]: [In here](https://github.com/mattermost/mattermost-webapp/blob/e1a412d5262e01ff4fad67a7b65b102791ddc8bd/components/search_results/search_results.jsx#L304), you may change it as:
```diff
@@ -290,7 +290,10 @@ class SearchResults extends React.Component {
         }
 
         return (
-            <div className='sidebar-right__body'>
+            <div
+                id='searchContainer'
+                className='sidebar-right__body'
+            >
                 <SearchResultsHeader>
                     {formattedTitle}
                 </SearchResultsHeader>
@@ -327,4 +330,4 @@ class SearchResults extends React.Component {
     }
 }
```

Comment (sij507) [reply e2e/cypress/integration/post_header/post_header_spec.js]: @saturninoabril I added the searchContainer ID to the file, but I am still getting the error message: "CypressError: Timed out retrying: Expected to find element: '#searchContainer', but never found it"
Screen Shot 2019-10-30 at 11 12 34 PM

Comment (sij507) [reply e2e/cypress/integration/post_header/post_header_spec.js]: @saturninoabril nvm, I figured out now. Thanks!

File: e2e/cypress/integration/post_header/post_header_spec.js (hunk 2 above)
Comment (sij507) [reply e2e/cypress/integration/post_header/post_header_spec.js]: @saturninoabril I have updated with changes, please review, thanks!

**Comment by jasonblais** [comment]
Looks like this one is ready for another review

**Comment by sij507** [comment]
@srkgupta When I run the test case on my local machine, it passed. Please see the screenshot.
Screen Shot 2019-11-18 at 10 22 21 AM

**Comment by saturninoabril** [comment]
@srkgupta @sij507 I think I found a workaround to stub and add delay to network request.  I'll push into this commit so you can test.

File: e2e/cypress/integration/post_header/post_header_spec.js [hunk 3]
```diff
@@ -217,4 +235,38 @@ describe('Post Header', () => {
             cy.get(`#SEARCH_flagIcon_${postId}`).siblings('.post__pinned-badge').should('exist');
         });
     });
+
+    it('M17442 Visual verification of "Searching" animation for Flagged and Pinned posts', () => {
+        // # Go to Town-Square channel
+        cy.visit('/ad-1/channels/town-square');
+
+        // Pin and flag ten posts before clicking on Pinned and Flagged post icons
+        pinAndFlagTenPosts();
```
Comment (saturninoabril) [inline e2e/cypress/integration/post_header/post_header_spec.js]: Please change above lines to:
```javascript
        cy.delayRequestToRoutes(['pinned', 'flagged'], 5000);
        cy.reload();

        // # Go to Town-Square channel
        cy.visit('/ad-1/channels/town-square');

        // Pin and flag ten posts before clicking on Pinned and Flagged post icons
        cy.postMessage('Post');

        //Pin and flag the posted message
        cy.getLastPostId().then((postId) => {
            cy.clickPostDotMenu(postId);
            cy.get(`#pin_post_${postId}`).click();
            cy.clickPostFlagIcon(postId);
        });
```

Then add this new Cypress command, preferably in the new file `cypress/support/fetch_commands.js`
```javascript
Cypress.Commands.add('delayRequestToRoutes', (routes = [], delay = 0) => {
    cy.on('window:before:load', (win) => addDelay(win, routes, delay));
});

const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const addDelay = (win, routes, delay) => {
    const fetch = win.fetch;
    cy.stub(win, 'fetch').callsFake((...args) => {
        for (let i = 0; i < routes.length; i++) {
            if (args[0].includes(routes[i])) {
                return wait(delay).then(() => fetch(...args));
            }
        }

        return fetch(...args);
    });
};
```

Be sure to import this new file at `e2e/cypress/support/index.js`

File: e2e/cypress/integration/post_header/post_header_spec.js [hunk 4]
```diff
@@ -8,6 +8,24 @@
 // - Use element ID when selecting an element. Create one if none.
 // ***************************************************************
 
+const pinnedFlaggedPosts = [];
+
+// Post 10 messages, pin and flag them, add the postIds to pinnedFlaggedPosts
+function pinAndFlagTenPosts() {
```
[the remark this answers is not in the document]
Comment (sij507) [reply e2e/cypress/integration/post_header/post_header_spec.js]: Removed

File: e2e/cypress/integration/post_header/post_header_spec.js (hunk 3 above)
Comment (sij507) [reply e2e/cypress/integration/post_header/post_header_spec.js]: Updated and added fetch_commands.js

**Comment by sij507** [comment]
@srkgupta @saturninoabril I updated with recommendations, it works on my local machine. Please review, thanks!
Screen Shot 2019-11-19 at 10 35 29 AM

**Pull request merged**